optimade/server/config.py

```diff
@@ -36,14 +35,74 @@ class LogLevel(Enum):
     CRITICAL = "critical"
 
 
+def config_file_settings(settings: BaseSettings) -> Dict[str, Any]:
+    """Configuration file settings source.
+
+    Based on the example in the
+    [pydantic documentation](https://pydantic-docs.helpmanual.io/usage/settings/#adding-sources),
+    this function loads ServerConfig settings from a configuration file.
+
+    The file must be of either type JSON or YML/YAML.
+
+    Parameters:
+        settings: The `pydantic.BaseSettings` class using this function as a
+            `pydantic.SettingsSourceCallable`.
+
+    Returns:
+        Dictionary of settings as read from a file.
+
+    """
+    import json
+    import os
+    import yaml
+
+    encoding = settings.__config__.env_file_encoding
+    config_file = Path(os.getenv("OPTIMADE_CONFIG_FILE", DEFAULT_CONFIG_FILE_PATH))
+
+    res = {}
+    if config_file.is_file():
+        config_file_content = config_file.read_text(encoding=encoding)
+
+        try:
+            res = json.loads(config_file_content)
+        except json.JSONDecodeError as json_exc:
+            try:
+                # This can essentially also load JSON files, as JSON is a subset of YAML v1,
+                # but I suspect it is not as rigorous
+                res = yaml.safe_load(config_file_content)
+            except yaml.YAMLError as yaml_exc:
+                warnings.warn(
+                    f"Unable to parse config file {config_file} as JSON and as YAML, using the "
+                    "default settings instead..\n"
+                    f"Errors:\n  JSON:\n{json_exc}.\n\n  YAML:\n{yaml_exc}"
+                )
+    else:
+        warnings.warn(
+            f"Unable to find config file at {config_file}, using the default settings instead."
+        )
+
+    if res:
+        # Set "config_file" parameter to acknowledge a config file has been used.
+        res["config_file"] = str(config_file)
+    elif res is None:
+        # This can happen if the yaml loading doesn't succeed properly.
+        res = {}
```

Comment:
Should we emit a warning here then? Does the yaml library really silently fail like this?

Reply:
It returns `None` if, e.g., the file content is empty.

Reply:
You were okay with not emitting a warning here, right?

Reply:
If an empty file is the only time this gets triggered then I don't have any strong feelings

Reply:
Cool. It's worth noting it would be any file with either none content or only whitespace content that this holds true for.

Reply:
I guess no-one would ever do this intentionally... hmmm